src/storage/hummock_sdk/src/change_log.rs

```diff
@@ -34,35 +32,70 @@ impl TableChangeLog {
         &self.0[start..end]
     }
+
+    pub fn truncate(&mut self, truncate_epoch: u64) {
+        // TODO: may optimize by using VecDeque to maintain the log
+        self.0
+            .retain(|change_log| *change_log.epochs.last().expect("non-empty") > truncate_epoch);
+        if let Some(first_log) = self.0.first_mut() {
+            first_log.epochs.retain(|epoch| *epoch > truncate_epoch);
+        }
+    }
 }
 
 impl TableChangeLog {
     pub fn to_protobuf(&self) -> PbTableChangeLog {
         PbTableChangeLog {
-            change_logs: self
-                .0
-                .iter()
-                .map(|epoch_new_log| PbEpochNewChangeLog {
-                    epochs: epoch_new_log.epochs.clone(),
-                    new_value: epoch_new_log.new_value.clone(),
-                    old_value: epoch_new_log.old_value.clone(),
-                })
-                .collect(),
+            change_logs: self.0.clone(),
         }
     }
 
     pub fn from_protobuf(val: &PbTableChangeLog) -> Self {
-        Self(
-            val.change_logs
-                .iter()
-                .map(|epoch_new_log| EpochNewChangeLog {
-                    epochs: epoch_new_log.epochs.clone(),
-                    new_value: epoch_new_log.new_value.clone(),
-                    old_value: epoch_new_log.old_value.clone(),
-                })
-                .collect(),
-        )
+        Self(val.change_logs.clone())
+    }
+}
+
+pub fn build_table_change_log_delta<'a>(
+    old_value_ssts: impl Iterator<Item = SstableInfo>,
+    new_value_ssts: impl Iterator<Item = &'a SstableInfo>,
+    epochs: &Vec<u64>,
+    log_store_table_ids: impl Iterator<Item = (u32, u64)>,
+) -> HashMap<TableId, ChangeLogDelta> {
+    let mut table_change_log: HashMap<_, _> = log_store_table_ids
+        .map(|(table_id, truncate_epoch)| {
+            (
+                TableId::new(table_id),
+                ChangeLogDelta {
+                    truncate_epoch,
+                    new_log: Some(EpochNewChangeLog {
+                        new_value: vec![],
+                        old_value: vec![],
+                        epochs: epochs.clone(),
+                    }),
+                },
+            )
+        })
+        .collect();
+    for sst in old_value_ssts {
+        for table_id in &sst.table_ids {
+            match table_change_log.get_mut(&TableId::new(*table_id)) {
+                Some(log) => {
+                    log.new_log.as_mut().unwrap().old_value.push(sst.clone());
+                }
+                None => {
+                    warn!(table_id, ?sst, "old value sst contains non-log-store table");
+                }
+            }
+        }
     }
+    for sst in new_value_ssts {
+        for table_id in &sst.table_ids {
+            if let Some(log) = table_change_log.get_mut(&TableId::new(*table_id)) {
+                log.new_log.as_mut().unwrap().new_value.push(sst.clone());
+            }
+        }
+    }
+    table_change_log
```

Comment:
Just thought of a corner case, will we generate a `old_value` sst if there are only INSERTs in `new_value` sst? Correct me if I am wrong, I think we will generate a `old_value` sst with no entries in it.

Reply:
By design, if there is no data, we will still have an empty `EpochNewChangeLog` to store the epochs information so that when we replay we won't miss a barrier.

I just checked the code. Currently the sst builder is initialized lazily. So if the payload is append only, the old value imm iterator won't have any output, and then no sst will be built.